modules/reindex/src/test/java/org/elasticsearch/index/reindex/ReindexFailoverIT.java

```diff
@@ -0,0 +1,125 @@
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class ReindexFailoverIT extends ReindexTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        final HashSet<Class<? extends Plugin>> classes = new HashSet<>(super.nodePlugins());
+        classes.add(MockTransportService.TestPlugin.class);
+        return classes;
+    }
+
+    public void testReindexFailover() throws Throwable {
+        logger.info("--> start 4 nodes, 1 master, 3 data");
+
+        final Settings sharedSettings = Settings.builder()
+            .put("cluster.join.timeout", "10s")  // still long to induce failures but not too long so test won't time out
+            .build();
+
+        internalCluster().startMasterOnlyNodes(1, sharedSettings);
+        internalCluster().startDataOnlyNodes(3, sharedSettings);
+
+        logger.info("--> wait for all nodes to join the cluster");
+        ensureStableCluster(4);
+
+        ClusterStateResponse clusterState = client().admin().cluster().prepareState().setNodes(true).get();
+        Map<String, String> nodeIdToName = new HashMap<>();
+        for (ObjectObjectCursor<String, DiscoveryNode> node : clusterState.getState().nodes().getDataNodes()) {
+            nodeIdToName.put(node.key, node.value.getName());
+        }
+
+        // Create Index
+        Settings.Builder indexSettings = Settings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 1);
+        client().admin().indices().prepareCreate("source").setSettings(indexSettings).get();
+        ensureGreen("source");
+
+        List<IndexRequestBuilder> docs = new ArrayList<>();
+        int docCount = between(1500, 5000);
+        for (int i = 0; i < docCount; i++) {
+            docs.add(client().prepareIndex("source", "_doc", Integer.toString(i)).setSource("foo", "a"));
+        }
+
+        indexRandom(true, docs);
+        assertHitCount(client().prepareSearch("source").setSize(0).get(), docCount);
+        ensureGreen("source");
+
+        // Copy all the docs
+        ReindexRequestBuilder copy = reindex().source("source").destination("dest").refresh(true);
+        ReindexRequest reindexRequest = copy.request();
+        reindexRequest.setScroll(TimeValue.timeValueSeconds(25));
+        StartReindexJobAction.Request request = new StartReindexJobAction.Request(reindexRequest, false);
+
+        copy.source().setSize(100);
+        StartReindexJobAction.Response response = client().execute(StartReindexJobAction.INSTANCE, request).get();
+        TaskId taskId = new TaskId(response.getTaskId());
+
+        String nodeId = taskId.getNodeId();
+        String nodeName = nodeIdToName.get(nodeId);
+
+        logger.info("--> restarting node: " + nodeName);
+        internalCluster().restartNode(nodeName, new InternalTestCluster.RestartCallback());
+
+        logger.info("--> waiting to cluster to heal");
+        ensureStableCluster(4);
```

Comment:
I don't think this is needed. restartNode will wait for the node to have rejoined the cluster.